From 29cfb62131ff6036d3c12b7c5fc938f52a2e7cf8 Mon Sep 17 00:00:00 2001
From: Pradyot Bhattacharyya <pbhattacharyya@habana.ai>
Date: Fri, 09 May 2025 19:32:42 +0300
Subject: [PATCH] [SW-224361]: add retry in case spi_transceive fails

SPI transceive fails at random during accessing tpm.
A immediate retry addresses this and unblocks the process

Change-Id: Ie0f47da4e6982ceec6e4122756412634da5f8434
Signed-off-by: Pradyot Bhattacharyya <pbhattacharyya@habana.ai>
---

diff --git a/drivers/tpm/tpm2_tis_spi.c b/drivers/tpm/tpm2_tis_spi.c
index 70312b3..b7adc73 100644
--- a/drivers/tpm/tpm2_tis_spi.c
+++ b/drivers/tpm/tpm2_tis_spi.c
@@ -48,6 +48,7 @@
 #define TPM_RID(l)			(0x0F04 | ((l) << 12))
 
 #define MAX_SPI_FRAMESIZE	32
+#define SPI_XFER_RETRY		5
 
 /* Number of wait states to wait for */
 #define TPM_WAIT_STATES		100
@@ -93,6 +94,7 @@
 	int transfer_len, ret, i;
 	uint8_t tx_bytes[MAX_SPI_FRAMESIZE];
 	uint8_t rx_bytes[MAX_SPI_FRAMESIZE];
+	uint8_t try = 0;
 	struct spi_buf tx_buf = {
 		.buf = tx_bytes,
 	};
@@ -168,8 +170,14 @@
 			memset(tx_bytes, 0, transfer_len);
 		}
 
-		ret = spi_transceive(data->spi_dev, &config->spi_cfg,
-				     &tx, in ? &rx : NULL);
+		while (try++ < SPI_XFER_RETRY) {
+			ret = spi_transceive(data->spi_dev, &config->spi_cfg,
+						&tx, in ? &rx : NULL);
+			if (!ret)
+				break;
+			LOG_INF("spi_transceive failed (%d)...retrying %d", ret, try);
+		}
+
 		if (ret) {
 			LOG_ERR("%s: spi read transfer failed (err: %d)\n",
 				__func__, ret);
